keylime/tpm2.py

```diff
@@ -1195,7 +1195,7 @@ def read_ekcert_nvram(self):
 
             # Read the RSA EK cert from NVRAM (DER format)
             if tools_version == "3.2":
-                retDict = self.__run("tpm2_nvread -x 0x1c00002 -s %s -f %s"%(ekcert_size, nvpath.name), raiseOnError=False, outputpaths=nvpath.name)
+                retDict = self.__run("tpm2_nvread -x 0x1c00002 -s %s -f %s -a 0x01c00002"%(ekcert_size, nvpath.name), raiseOnError=False, outputpaths=nvpath.name)
```

Comment:
Should `-a` be set to `0x40000001`?   That should be owner auth (https://github.com/tpm2-software/tpm2-tools/blob/3.X/man/tpm2_nvread.1.md)

Reply:
Hmm, that change gives me an exception:

```
2019-11-18 16:45:27.676 - keylime.tpm2 - INFO - Flushing old ek handle: 0x81010007
Traceback (most recent call last):
  File "/usr/local/bin/keylime_agent", line 11, in <module>
    load_entry_point('keylime==1.2', 'console_scripts', 'keylime_agent')()
  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/cloud_agent.py", line 491, in main
    (ek,ekcert,aik,ek_tpm,aik_name) = tpm.tpm_init(self_activate=False,config_pw=config.get('cloud_agent','tpm_ownerpassword')) # this tells initialize not to self activate the AIK
  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/tpm2.py", line 903, in tpm_init
    ekcert = self.read_ekcert_nvram()
  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/tpm2.py", line 1207, in read_ekcert_nvram
    raise Exception("tpm2_nvread for ekcert failed with code "+str(code)+": "+str(errout))
Exception: tpm2_nvread for ekcert failed with code 1: ['ERROR: Failed to read NVRAM area at index 0x1c00002 (29360130). Error:0x9a2\n', 'ERROR: Unable to run tpm2_nvread\n']
```

![](https://i.imgur.com/uhGbkLE.png)

Reply:
You might need to pass in the owner password now via `-P` like we did for reading the key from nvram, though this shouldn't be necessary for `0x1c00002` (this is supposed to be a publicly-available EK cert). 

Are you sure this functionality has changed, and it isn't just something specific to your TPM chip?  This code hasn't been updated in over 6 months (and the change isn't related to this). 
 
https://github.com/tpm2-software/tpm2-tools/blob/3.X/tools/tpm2_nvread.c

Does your TPM not have its EK cert stored at `0x1c00002`?

Reply:
Maybe this is also related to this potential bug in tpm2-abrmd?  

https://github.com/tpm2-software/tpm2-abrmd/issues/658

Reply:
@jetwhiz most like is from the looks of it.

I will put an agenda item on to discuss the tpm tooling we use. I am spending most of my time now trying to debug tpm2-tools not working or make changes as arguments are changed or commands renamed etc. I want to be getting on with working on core features, but I am spending too much time just keeping things working (which means taking your time too as the SME in this area).

I would like to propose we evaluate https://github.com/google/go-tpm , I spoke to some of the developers on there and they are very open to any commands not being present, getting added. I have feeling this will be less of a moving target perhaps..at the moment we are dealing with the impact of changes in three projects , tpm2-tools, tpm2-abrmd, tpm2-tss and its causing a lot of uneeded grief. We should be able to focus on just working on our application, not fixing an every changing and frequently breaking stack.

Please take a look at go-tpm at let me know your thoughts, I can see there is no resource manager, but I wonder if that matters to us?